session/nontransactional_test.go

```diff
@@ -0,0 +1,149 @@
+			tk.MustExec(fmt.Sprintf("insert into t values ('%d', %d)", i, i*2))
+		}
+		if strings.Contains(table, "a int") {
+			rows := tk.MustQuery("split on a limit 3 dry run delete from t").Rows()
+			for _, row := range rows {
+				require.True(t, strings.HasPrefix(row[0].(string), "DELETE FROM `test`.`t` WHERE `a` BETWEEN"))
+			}
+		}
+		tk.MustQuery("split on a limit 3 dry run query delete from t").Check(testkit.Rows(
+			"select `a` from `test`.`t` where true order by IF(ISNULL(`a`),0,1),`a`"))
+	}
+}
+
+func TestNonTransactionalDeleteErrorMessage(t *testing.T) {
+	store, clean := createStorage(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set @@tidb_max_chunk_size=35")
+	tk.MustExec("use test")
+	tk.MustExec("create table t(a int, b int, primary key(a, b) clustered)")
+	for i := 0; i < 100; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values ('%d', %d)", i, i*2))
+	}
+	failpoint.Enable("github.com/pingcap/tidb/session/splitDeleteError", `return`)
+	defer failpoint.Disable("github.com/pingcap/tidb/session/splitDeleteError")
+	rows := tk.MustQuery("split on a limit 3 delete from t").Rows()
+	require.Equal(t, 1, len(rows))
+	require.Equal(t, rows[0][2].(string), "Early return: error occurred in the first job: injected split delete error")
+}
+
+func TestNonTransactionalDeleteSplitOnTiDBRowID(t *testing.T) {
+	store, clean := createStorage(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set @@tidb_max_chunk_size=35")
+	tk.MustExec("use test")
+	tk.MustExec("create table t(a int, b int)")
+	for i := 0; i < 100; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values ('%d', %d)", i, i*2))
+	}
+	tk.MustExec("split on _tidb_rowid limit 3 delete from t")
+	tk.MustQuery("select count(*) from t").Check(testkit.Rows("0"))
+}
+
+func TestNonTransactionalDeleteNull(t *testing.T) {
+	store, clean := createStorage(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set @@tidb_max_chunk_size=35")
+	tk.MustExec("use test")
+	tk.MustExec("create table t(a int, b int, key(a))")
+	for i := 0; i < 100; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values ('%d', %d)", i, i*2))
+		tk.MustExec("insert into t values (null, null)")
+	}
+
+	tk.MustExec("split on a limit 3 delete from t")
+	tk.MustQuery("select count(*) from t").Check(testkit.Rows("0"))
+
+	// all values are null
+	for i := 0; i < 100; i++ {
+		tk.MustExec("insert into t values (null, null)")
+	}
+	tk.MustExec("split on a limit 3 delete from t")
+	tk.MustQuery("select count(*) from t").Check(testkit.Rows("0"))
+}
+
+func TestNonTransactionalDeleteSmallBatch(t *testing.T) {
+	store, clean := createStorage(t)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("set @@tidb_max_chunk_size=1024")
+	tk.MustExec("use test")
+	tk.MustExec("create table t(a int, b int, key(a))")
+	for i := 0; i < 10; i++ {
+		tk.MustExec(fmt.Sprintf("insert into t values ('%d', %d)", i, i*2))
+		tk.MustExec("insert into t values (null, null)")
```

Comment:
If there are too many null values in column `a`, will the job with `a is null` got "txn too large" failed?

Reply:
That's possible if there are duplicated values, not limited to NULLs.
So we would suggest users split on a column indexed by a unique index.